share/man/man4/le.4
.\"	$NetBSD: le.4,v 1.22 2004/10/04 19:12:52 rumble Exp $
.\"
.\"-
.\" Copyright (c) 1992, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This software was developed by the Computer Systems Engineering group
.\" at Lawrence Berkeley Laboratory under DARPA contract BG 91-66 and
.\" contributed to Berkeley.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	from: Header: le.4,v 1.2 92/10/13 05:31:33 leres Exp
.\"	from: @(#)le.4	8.1 (Berkeley) 6/9/93
.\" $FreeBSD$
.\"
.Dd February 15, 2017
.Dt LE 4
.Os
.Sh NAME
.Nm le
.Nd "AMD Am7900 LANCE and Am79C9xx ILACC/PCnet Ethernet interface driver"
.Sh SYNOPSIS
To compile this driver into the kernel,
place the following line in your
kernel configuration file:
.Bd -ragged -offset indent
.Cd "device le"
.Ed
.Pp
Alternatively, to load the driver as a
module at boot time, place the following line in
.Xr loader.conf 5 :
.Bd -literal -offset indent
if_le_load="YES"
.Ed
.Pp
For ISA non-PnP adapters, the port address as well as the IRQ and the DRQ
numbers have to be specified in
.Pa /boot/device.hints :
.Cd hint.le.0.at="isa"
.Cd hint.le.0.port="0x280"
.Cd hint.le.0.irq="10"
.Cd hint.le.0.drq="0"
.Sh DESCRIPTION
The
.Nm
driver provides support for Ethernet adapters based on the
.Tn AMD Am7990
and
.Tn Am79C90
.Pq CMOS, pin-compatible
Local Area Network Controller for Ethernet
.Pq Tn LANCE
chips.
.Pp
The
.Nm
driver also supports Ethernet adapters based on the
.Tn AMD Am79C900
Integrated Local Area Communications Controller
.Pq Tn ILACC
as well as the
.Tn Am79C9xx PCnet
family of chips, which are single-chip implementations of a
.Tn LANCE
chip and a DMA engine.
This includes a superset of the
.Tn PCI
bus Ethernet chips supported by the
.Xr pcn 4
driver.
The
.Nm
driver treats all of these
.Tn PCI
bus Ethernet chips as an
.Tn AMD Am79C970 PCnet-PCI
and does not support the additional features like the MII bus and burst mode of
.Tn AMD Am79C971 PCnet-FAST
and greater chips.
Thus the
.Xr pcn 4
driver should be preferred for the latter.
.Pp
Generally, the
.Nm
driver aims at supporting as many different chips on as many different
platforms as possible,
partially at the cost of the best performance with some of these.
.Pp
The
.Nm
driver supports reception and transmission of extended frames for
.Xr vlan 4 .
Selective reception of multicast Ethernet frames is provided by a 64-bit mask;
multicast destination addresses are hashed to a bit entry using the Ethernet
CRC function.
.Sh HARDWARE
.Ss ISA
The
.Nm
driver supports
.Tn ISA
bus Ethernet adapters which are based on the following chips:
.Pp
.Bl -bullet -compact
.It
.Tn AMD Am7990 and Am79C90 LANCE
.It
.Tn AMD Am79C960 PCnet-ISA
.It
.Tn AMD Am79C961 PCnet-ISA+
.It
.Tn AMD Am79C961A PCnet-ISA II
.El
.Pp
This includes support for the following Ethernet adapters:
.Pp
ISA non-PnP:
.Pp
.Bl -bullet -compact
.It
.Tn BICC Isolan
.\" .It
.\" .Tn Digital DEPCA
.It
.Tn Novell NE2100
.El
.Pp
ISA PnP:
.Pp
.Bl -bullet -compact
.It
.Tn AMD AM1500T/AM2100
.It
.Tn AMD PCnet-32
.It
.Tn AMD PCnet-ISA
.It
.Tn Allied Telesyn AT-1500
.It
.Tn Boca LANCard Combo
.It
.Tn Cabletron E2100 Series DNI
.It
.Tn Cabletron E2200 Single Chip
.It
.Tn Melco Inc. LGY-IV
.It
.Tn Novell NE2100
.It
.Tn Racal InterLan EtherBlaster
.El
.Pp
The
.Nm
driver does not support the selection of media types and options via
.Xr ifconfig 8
with
.Tn ISA
bus Ethernet adapters.
.Ss PCI
The
.Tn PCI
bus Ethernet chips supported by the
.Nm
driver are:
.Pp
.Bl -bullet -compact
.It
.Tn AMD Am53C974/Am79C970/Am79C974 PCnet-PCI
.It
.Tn AMD Am79C970A PCnet-PCI II
.It
.Tn AMD Am79C971 PCnet-FAST
.It
.Tn AMD Am79C972 PCnet-FAST+
.It
.Tn AMD Am79C973/Am79C975 PCnet-FAST III
.It
.Tn AMD Am79C976 PCnet-PRO
.It
.Tn AMD Am79C978 PCnet-Home
.El
.Pp
This includes support for the following Ethernet adapters:
.Pp
.Bl -bullet -compact
.It
.Tn AcerLAN NIC P20
.It
.Tn Allied Telesyn AT-2450 and AT-2700 series
.It
.Tn VMware emulated AMD Am79C970A PCnet-PCI II interface
.El
.Pp
The
.Nm
driver supports the selection of the following media types via
.Xr ifconfig 8
with
.Tn PCI
bus Ethernet adapters:
.Bl -tag -width ".Cm 10base5/AUI"
.It Cm autoselect
Enable autoselection of the media type.
.It Cm 10baseT/UTP
Select UTP media.
.It Cm 10base5/AUI
Select AUI/BNC media.
.El
.Pp
The following media option is supported with these media types:
.Bl -tag -width ".Cm full-duplex"
.It Cm full-duplex
Select full duplex operation.
.El
.Pp
Note that unlike the
.Xr pcn 4
driver, the
.Nm
driver does not support selecting 100Mbps (Fast Ethernet) media types.
.Ss sparc64
The
.Nm
driver supports the on-board
.Tn LANCE
interfaces found in
.Tn Sun Ultra 1
machines.
The
.Nm
driver allows the selection of the following media types via
.Xr ifconfig 8
with these on-board interfaces:
.Bl -tag -width ".Cm 10base5/AUI"
.It Cm autoselect
Enable autoselection of the media type.
.It Cm 10baseT/UTP
Select UTP media.
.It Cm 10base5/AUI
Select AUI media.
.El
.Pp
When using autoselection, a default media type is selected for use by
examining all ports for carrier.
The first media type with which a carrier is detected will be selected.
Additionally, if carrier is dropped on a port, the driver will switch
between the possible ports until one with carrier is found.
.Pp
The
.Nm
driver also supports the following
.Tn Sun SBus
Ethernet add-on adapters:
.Pp
.Bl -bullet -compact
.It
.Tn SCSI HBA and Buffered Ethernet
.Pq SBE/S, P/N 501-1869
.It
.Tn Fast SCSI and Buffered Ethernet
.Pq FSBE/S, P/N 501-2015 and 501-2981
.El
.Pp
The
.Nm
driver does not support the selection of media types and options via
.Xr ifconfig 8
with
.Tn SBus
Ethernet add-on adapters.
.Pp
For further information on configuring media types and options, see
.Xr ifconfig 8 .
.Sh DIAGNOSTICS
.Bl -diag
.It "le%d: overflow"
More packets came in from the Ethernet than there was space in the
.Tn LANCE
receive buffers.
Packets were missed.
.It "le%d: receive buffer error"
The
.Tn LANCE
ran out of buffer space, packet dropped.
.It "le%d: lost carrier"
The Ethernet carrier disappeared during an attempt to transmit.
The
.Tn LANCE
will finish transmitting the current packet,
but will not automatically retry transmission if there is a collision.
.It "le%d: excessive collisions, tdr %d"
The Ethernet was extremely busy or jammed,
outbound packets were dropped after 16 attempts to retransmit.
.Pp
TDR
is the abbreviation of
.Qq Time Domain Reflectometry .
The optionally reported TDR value is an internal counter of the interval
between the start of a transmission and the occurrence of a collision.
This value can be used to determine the distance from the Ethernet tap to
the point on the Ethernet cable that is shorted or open (unterminated).
.It "le%d: dropping chained buffer"
A packet did not fit into a single receive buffer and was dropped.
Since the
.Nm
driver allocates buffers large enough to receive maximum sized Ethernet
packets, this means some other station on the LAN transmitted a packet
larger than allowed by the Ethernet standard.
.It "le%d: transmit buffer error"
The
.Tn LANCE
ran out of buffer space before finishing the transmission of a packet.
If this error occurs, the driver software has a bug.
.It "le%d: underflow"
The
.Tn LANCE
ran out of buffer space before finishing the transmission of a packet.
If this error occurs, the driver software has a bug.
.It "le%d: controller failed to initialize"
Driver failed to start the
.Tn LANCE .
This is potentially a hardware failure.
.It "le%d: memory error"
RAM failed to respond within the timeout when the
.Tn LANCE
wanted to read or write it.
This is potentially a hardware failure.
.It "le%d: receiver disabled"
The receiver of the
.Tn LANCE
was turned off due to an error.
.It "le%d: transmitter disabled"
The transmitter of the
.Tn LANCE
was turned off due to an error.
.El
.Sh SEE ALSO
.Xr altq 4 ,
.Xr arp 4 ,
.Xr intro 4 ,
.Xr netintro 4 ,
.Xr pcn 4 ,
.Xr vlan 4 ,
.Xr ifconfig 8
.Sh HISTORY
The
.Nm
driver was ported from
.Nx
and first appeared in
.Fx 6.1 .
The
.Nx
version in turn was derived from the
.Nm
driver which first appeared in
.Bx 4.4 .
.Sh AUTHORS
The
.Nm
driver was ported by
.An Marius Strobl Aq Mt marius@FreeBSD.org .
.\" .Sh BUGS
.\" The Am7990 Revision C chips have a bug which causes garbage to be inserted
.\" in front of the received packet occasionally.
.\" The work-around is to ignore packets with an invalid destination address
.\" (garbage will usually not match), by double-checking the destination
.\" address of every packet in the driver.
.\" This work-around can be enabled with the
.\" .Dv LANCE_REVC_BUG
.\" kernel option.
.\" .Pp
.\" When
.\" .Dv LANCE_REVC_BUG
.\" is enabled, the
.\" .Nm
.\" driver executes one or two calls to an inline Ethernet address comparison
.\" function for every received packet.
.\" On the
.\" .Tn MC68000
.\" it is exactly eight instructions of 16 bits each.
.\" There is one comparison for each unicast packet, and two comparisons for
.\" each broadcast packet.
.\" .Pp
.\" In summary, the cost of the LANCE_REVC_BUG option is:
.\" .Bl -enum -compact
.\" .It
.\" loss of multicast support, and
.\" .It
.\" eight extra
.\" .Tn CPU
.\" instructions per received packet, sometimes sixteen, depending on both the
.\" processor, and the type of packet.
.\" .El
.\" .Pp
.\" All sun3 systems are presumed to have this bad revision of the Am7990,
.\" until proven otherwise.
.\" Alas, the only way to prove what revision of the chip is in a particular
.\" system is inspection of the date code on the chip package,
.\" to compare against a list of what chip revisions were fabricated between
.\" which dates.
.\" .Pp
.\" Alas, the Am7990 chip is so old that
.\" .Tn AMD
.\" has
.\" .Qq de-archived
.\" the production information about it; pending a search elsewhere, we do not
.\" know how to identify the revision C chip from the date codes.


usr.sbin/bluetooth/hccontrol/le.c
/*
 * le.c
 *
 * Copyright (c) 2015 Takanori Watanabe <takawata@freebsd.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: hccontrol.c,v 1.5 2003/09/05 00:38:24 max Exp $
 * $FreeBSD$
 */

#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/sysctl.h>
#include <sys/select.h>
#include <assert.h>
#include <bitstring.h>
#include <err.h>
#include <errno.h>
#include <netgraph/ng_message.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdint.h>
#define L2CAP_SOCKET_CHECKED
#include <bluetooth.h>
#include "hccontrol.h"

static int le_set_scan_param(int s, int argc, char *argv[]);
static int le_set_scan_enable(int s, int argc, char *argv[]);
static int parse_param(int argc, char *argv[], char *buf, int *len);
static int le_set_scan_response(int s, int argc, char *argv[]);
static int le_read_supported_status(int s, int argc, char *argv[]);
static int le_read_local_supported_features(int s, int argc ,char *argv[]);
static int set_le_event_mask(int s, uint64_t mask);
static int set_event_mask(int s, uint64_t mask);
static int le_enable(int s, int argc, char *argv[]);

static int
le_set_scan_param(int s, int argc, char *argv[])
{
	int type;
	int interval;
	int window;
	int adrtype;
	int policy;
	int e, n;

	ng_hci_le_set_scan_parameters_cp cp;
	ng_hci_le_set_scan_parameters_rp rp;

	if (argc != 5)
		return USAGE;
	
	if (strcmp(argv[0], "active") == 0)
		type = 1;
	else if (strcmp(argv[0], "passive") == 0)
		type = 0;
	else
		return USAGE;

	interval = (int)(atof(argv[1])/0.625);
	interval = (interval < 4)? 4: interval;
	window = (int)(atof(argv[2])/0.625);
	window = (window < 4) ? 4 : interval;
	
	if (strcmp(argv[3], "public") == 0)
		adrtype = 0;
	else if (strcmp(argv[3], "random") == 0)
		adrtype = 1;
	else
		return USAGE;

	if (strcmp(argv[4], "all") == 0)
		policy = 0;
	else if (strcmp(argv[4], "whitelist") == 0)
		policy = 1;
	else
		return USAGE;

	cp.le_scan_type = type;
	cp.le_scan_interval = interval;
	cp.own_address_type = adrtype;
	cp.le_scan_window = window;
	cp.scanning_filter_policy = policy;
	n = sizeof(rp);
	e = hci_request(s, NG_HCI_OPCODE(NG_HCI_OGF_LE,
		NG_HCI_OCF_LE_SET_SCAN_PARAMETERS), 
		(void *)&cp, sizeof(cp), (void *)&rp, &n);

	return 0;
}

static int
le_set_scan_enable(int s, int argc, char *argv[])
{
	ng_hci_le_set_scan_enable_cp cp;
	ng_hci_le_set_scan_enable_rp rp;
	int e, n, enable = 0;

	if (argc != 1)
		return USAGE;
	  
	if (strcmp(argv[0], "enable") == 0)
		enable = 1;
	else if (strcmp(argv[0], "disable") != 0)
		return USAGE;

	n = sizeof(rp);
	cp.le_scan_enable = enable;
	cp.filter_duplicates = 0;
	e = hci_request(s, NG_HCI_OPCODE(NG_HCI_OGF_LE,
		NG_HCI_OCF_LE_SET_SCAN_ENABLE), 
		(void *)&cp, sizeof(cp), (void *)&rp, &n);
			
	if (e != 0 || rp.status != 0)
		return ERROR;

	return OK;
}

static int
parse_param(int argc, char *argv[], char *buf, int *len)
{
	char *buflast  =  buf + (*len);
	char *curbuf = buf;
	char *token,*lenpos;
	int ch;
	int datalen;
	uint16_t value;
	optreset = 1;
	optind = 0;
	while ((ch = getopt(argc, argv , "n:f:u:")) != -1) {
		switch(ch){
		case 'n':
			datalen = strlen(optarg);
			if ((curbuf + datalen + 2) >= buflast)
				goto done;
			curbuf[0] = datalen + 1;
			curbuf[1] = 8;
			curbuf += 2;
			memcpy(curbuf, optarg, datalen);
			curbuf += datalen;
			break;
		case 'f':
			if (curbuf+3 > buflast)
				goto done;
			curbuf[0] = 2;
			curbuf[1] = 1;
			curbuf[2] = atoi(optarg);
			curbuf += 3;
			break;
		case 'u':
			lenpos = buf;
			if ((buf+2) >= buflast)
				goto done;
			curbuf[1] = 2;
			*lenpos = 1;
			curbuf += 2;
			while ((token = strsep(&optarg, ",")) != NULL) {
				value = strtol(token, NULL, 16);
				if ((curbuf+2) >= buflast)
					break;
				curbuf[0] = value &0xff;
				curbuf[1] = (value>>8)&0xff;
				curbuf += 2;
			}
				
		}
	}
done:
	*len = curbuf - buf;

	return OK;
}

static int
le_set_scan_response(int s, int argc, char *argv[])
{
	ng_hci_le_set_scan_response_data_cp cp;
	ng_hci_le_set_scan_response_data_rp rp;
	int n;
	int e;
	int len;
	char buf[NG_HCI_ADVERTISING_DATA_SIZE];

	len = sizeof(buf);
	parse_param(argc, argv, buf, &len);
	memset(cp.scan_response_data, 0, sizeof(cp.scan_response_data));
	cp.scan_response_data_length = len;
	memcpy(cp.scan_response_data, buf, len);
	n = sizeof(rp);
	e = hci_request(s, NG_HCI_OPCODE(NG_HCI_OGF_LE,
			NG_HCI_OCF_LE_SET_SCAN_RESPONSE_DATA), 
			(void *)&cp, sizeof(cp), (void *)&rp, &n);
			
	printf("SET SCAN RESPONSE %d %d %d\n", e, rp.status, n);

	return OK;
}

static int
le_read_local_supported_features(int s, int argc ,char *argv[])
{
	ng_hci_le_read_local_supported_features_rp rp;
	int e;
	int n = sizeof(rp);

	e = hci_simple_request(s,
			NG_HCI_OPCODE(NG_HCI_OGF_LE,
			NG_HCI_OCF_LE_READ_LOCAL_SUPPORTED_FEATURES), 
			(void *)&rp, &n);

	printf("LOCAL SUPPORTED: %d %d %jx\n", e, rp.status,
	       (uintmax_t) rp.le_features);

	return 0;
}

static int
le_read_supported_status(int s, int argc, char *argv[])
{
	ng_hci_le_read_supported_status_rp rp;
	int e;
	int n = sizeof(rp);

	e = hci_simple_request(s, NG_HCI_OPCODE(
					NG_HCI_OGF_LE,
					NG_HCI_OCF_LE_READ_SUPPORTED_STATUS),
			       		(void *)&rp, &n);

	printf("LE_STATUS: %d %d %jx\n", e, rp.status, (uintmax_t)rp.le_status);

	return 0;
}

static int
set_le_event_mask(int s, uint64_t mask)
{
	ng_hci_le_set_event_mask_cp semc;
	ng_hci_le_set_event_mask_rp rp;  
	int i, n ,e;
	
	n = sizeof(rp);
	
	for (i=0; i < NG_HCI_LE_EVENT_MASK_SIZE; i++) {
		semc.event_mask[i] = mask&0xff;
		mask >>= 8;
	}
	e = hci_request(s, NG_HCI_OPCODE(NG_HCI_OGF_LE,
			NG_HCI_OCF_LE_SET_EVENT_MASK),
			(void *)&semc, sizeof(semc), (void *)&rp, &n);
	
	return 0;
}

static int
set_event_mask(int s, uint64_t mask)
{
	ng_hci_set_event_mask_cp semc;
	ng_hci_set_event_mask_rp rp;  
	int i, n, e;
	
	n = sizeof(rp);
	
	for (i=0; i < NG_HCI_EVENT_MASK_SIZE; i++) {
		semc.event_mask[i] = mask&0xff;
		mask >>= 8;
	}
	e = hci_request(s, NG_HCI_OPCODE(NG_HCI_OGF_HC_BASEBAND,
			NG_HCI_OCF_SET_EVENT_MASK),
			(void *)&semc, sizeof(semc), (void *)&rp, &n);
	
	return 0;
}

static
int le_enable(int s, int argc, char *argv[])
{
	if (argc != 1)
		return USAGE;
	
	if (strcasecmp(argv[0], "enable") == 0) {
		set_event_mask(s, NG_HCI_EVENT_MASK_DEFAULT |
			       NG_HCI_EVENT_MASK_LE);
		set_le_event_mask(s, NG_HCI_LE_EVENT_MASK_ALL);
	} else if (strcasecmp(argv[0], "disble") == 0)
		set_event_mask(s, NG_HCI_EVENT_MASK_DEFAULT);
	else
		return USAGE;

	return OK;
}

struct hci_command le_commands[] = {
{
	"le_enable",
	"le_enable [enable|disable] \n"
	"Enable LE event ",
	&le_enable,
},
  {
	  "le_read_local_supported_features",
	  "le_read_local_supported_features\n" 
	  "read local supported features mask",
	  &le_read_local_supported_features,
  },
  {
	  "le_read_supported_status",
	  "le_read_supported_status\n"
	  "read supported status"	  
	  ,
	  &le_read_supported_status,
  },
  {
	  "le_set_scan_response",
	  "le_set_scan_response -n $name -f $flag -u $uuid16,$uuid16 \n"
	  "set LE scan response data"
	  ,
	  &le_set_scan_response,
  },
  {
	  "le_set_scan_enable",
	  "le_set_scan_enable [enable|disable] \n"
	  "enable or disable LE device scan",
	  &le_set_scan_enable
  },
  {
	  "le_set_scan_param",
	  "le_set_scan_param [active|passive] interval(ms) window(ms) [public|random] [all|whitelist] \n"
	  "set LE device scan parameter",
	  &le_set_scan_param
  },
};
